sys/drv/rtl8139.c

```diff
@@ -32,18 +37,72 @@ static int rtl_reset(rtl8139_state_t *state) {
   return 1;
 }
 
-static int rtl8139_attach(device_t *dev) {
+static intr_filter_t rtl8139_intr(void *data) {
+  rtl8139_state_t *state = data;
+  // TODO: add enqueue routines
+  bus_write_2(state->regs, RL_ISR, RL_ISR_RX_OK);
+  return IF_FILTERED;
+}
+
+static int set_mem_contig_address(rtl8139_state_t *state) {
+  paddr_t page2; /* address of second allocated page */
+
+  /* set and check if any address is assigned */
+  if (!pmap_kextract(state->rx_buf, &state->paddr))
+    return -1;
+
+  /* check if memory is contiguous */
+  if (!pmap_kextract(state->rx_buf + PAGESIZE, &page2))
+    return -1;
+
+  if (state->paddr + PAGESIZE != page2)
+    return -1;
+
+  return 0;
+}
+
+int rtl8139_attach(device_t *dev) {
   rtl8139_state_t *state = dev->state;
 
+  pci_enable_busmaster(dev);
   state->regs = device_take_memory(dev, 1, RF_ACTIVE);
-  if (state->regs == NULL)
+
+  /*
+   * XXX: there is no irq routing in mimiker
+   *      so in this case we need to add irq manually
+   */
+  device_add_irq(dev, 0, 10);
+  state->irq_res = device_take_irq(dev, 0, RF_ACTIVE);
+  if (!state->regs || !state->irq_res) {
+    klog("rtl8139 failed to init resources");
     return ENXIO;
+  }
+  bus_intr_setup(dev, state->irq_res, rtl8139_intr, NULL, state, "RTL8139");
+
+  // TODO: mark this memory as PMAP_NOCACHE
+  state->rx_buf = (vaddr_t)kmem_alloc(2 * PAGESIZE, M_ZERO);
```

Comment:
Something like `RX_BUF_SIZE` is more readable than `2 * PAGESIZE` for me.

Reply:
Done